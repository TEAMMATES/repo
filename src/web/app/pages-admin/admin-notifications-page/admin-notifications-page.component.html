<div>
  <button id="btn-add-session" class="btn btn-success" (click)="isNotificationEditFormExpanded = true"
          [disabled]="isNotificationEditFormExpanded || isNotificationLoading"><i class="fas fa-plus"></i> Add New Notification
  </button>
</div>

<div class="margin-top-30px" *ngIf="isNotificationEditFormExpanded && !isNotificationLoading" @collapseAnim>
  <tm-notification-edit-form [formMode]="NotificationEditFormMode.ADD" [(model)]="notificationEditFormModel"
                        (addNewSessionEvent)="addNewNotificationHandler()"
                        (closeEditFormEvent)="isNotificationEditFormExpanded = false"
  ></tm-notification-edit-form>
</div>

<tm-loading-retry [shouldShowRetry]="hasNotificationLoadingFailed" [message]="'Failed to load data'"
                  (retryEvent)="retryLoadingAllData()">
  <div class="margin-top-30px">
    <div *tmIsLoading="isNotificationLoading">
      <div class="alert alert-warning margin-top-30px" role="alert" *ngIf="notifications.length === 0 && !isNotificationLoading">
        You have not created any notifications yet. Use the form above to create a notification.
      </div>
      
      <!-- Table Here -->
    </div>
  </div>
  <br/>
</tm-loading-retry>