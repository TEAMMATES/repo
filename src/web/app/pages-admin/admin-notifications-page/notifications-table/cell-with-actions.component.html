<div>
    <ng-template #actionButton let-isEnabled="isEnabled" let-tooltip="tooltip" let-name="name" let-id="id"
             let-tmRouterLink="tmRouterLink" let-queryParams="queryParams">
        <a class="btn btn-light btn-sm btn-margin-right custom-button" target="_blank" rel="noopener noreferrer"
            [id]="id"
            [ngClass]="{'disabled mouse-hover-only': !isEnabled}"
            [ngbTooltip]="isEnabled ? tooltip : 'You do not have the permissions to access this feature'"
            [tmRouterLink]="tmRouterLink"
            [queryParams]="queryParams">
        {{name}}
        </a>
    </ng-template>
    <ng-container>
      <button type="button" class="btn btn-light btn-sm" (click)="editNotification()" style="margin-right: 6px;">
            Edit
      </button>
      <button type="button" class="btn btn-danger btn-sm" (click)="deleteNotification()">
        Delete
      </button>
    </ng-container>

</div>
