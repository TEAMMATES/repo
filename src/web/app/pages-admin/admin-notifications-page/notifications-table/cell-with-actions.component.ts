import { CommonModule } from '@angular/common';
import { Component, Input } from '@angular/core';
import {
  NgbDropdownModule,
  NgbTooltipModule,
} from '@ng-bootstrap/ng-bootstrap';
import { AjaxLoadingModule } from '../../../components/ajax-loading/ajax-loading.module';
import { TeammatesRouterModule } from '../../../components/teammates-router/teammates-router.module';

@Component({
  selector: 'tm-group-buttons',
  templateUrl: './cell-with-actions.component.html',
  standalone: true,
  imports: [
    CommonModule,
    TeammatesRouterModule,
    AjaxLoadingModule,
    NgbDropdownModule,
    NgbTooltipModule,
  ],
})

export class CellWithActionsComponent {
  @Input() idx: number = 0;
  @Input() notificationId: number = 0;

  @Input() editNotification: () => void = () => {};
  @Input() deleteNotification : () => void = () => {};

}
