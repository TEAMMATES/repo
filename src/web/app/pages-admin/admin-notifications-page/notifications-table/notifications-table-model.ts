// FIXME: This Notification is to be imported from api-output after GET route PR is merged
import { Notification } from '../../../../services/notification.service';

/**
 * The model for a row in the notifications table.
 */
export interface NotificationsTableRowModel {
  index: number,
  notification: Notification,
}

/**
 * The color scheme of the header of the table
 */
export enum NotificationsTableHeaderColorScheme {
  /**
   * Blue background with white text.
   */
  BLUE,

  /**
   * White background with black text.
   */
  WHITE,
}
