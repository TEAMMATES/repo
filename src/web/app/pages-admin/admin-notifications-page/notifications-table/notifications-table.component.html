<div>
  <tm-sortable-table
    [columns]="columnsData"
    [rows]="rowsData"
    [initialSortBy]="tableSortBy"
    [sortOrder]="tableSortOrder"
    [headerColorScheme]="headerColorScheme"
    (sortEvent)="sortNotificationsTableRowModels($event)">
</tm-sortable-table>
</div>