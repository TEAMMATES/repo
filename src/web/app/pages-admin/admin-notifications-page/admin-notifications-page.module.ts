import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';
import { AdminNotificationsPageComponent } from './admin-notifications-page.component';

const routes: Routes = [
  {
    path: '',
    component: AdminNotificationsPageComponent,
  },
];

/**
 * Module for admin timezone page.
 */
@NgModule({
  declarations: [
    AdminNotificationsPageComponent,
  ],
  exports: [
    AdminNotificationsPageComponent,
  ],
  imports: [
    CommonModule,
    RouterModule.forChild(routes),
    LoadingSpinnerModule,
  ],
})
export class AdminNotificationsPageModule { }
