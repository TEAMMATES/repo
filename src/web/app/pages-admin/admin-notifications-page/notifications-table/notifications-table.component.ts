import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { CellWithActionsComponent } from './cell-with-actions.component';
import { NotificationsTableHeaderColorScheme, NotificationsTableRowModel } from './notifications-table-model';
import { SimpleModalService } from '../../../../services/simple-modal.service';
import { TimezoneService } from '../../../../services/timezone.service';
import { Notification } from '../../../../types/api-output';
import { SortBy, SortOrder } from '../../../../types/sort-properties';
import { SimpleModalType } from '../../../components/simple-modal/simple-modal-type';
import {
  ColumnData,
  SortableEvent,
  SortableTableCellData,
  SortableTableHeaderColorScheme,
} from '../../../components/sortable-table/sortable-table.component';
import { FormatDateBriefPipe } from '../../../components/teammates-common/format-date-brief.pipe';
import { FormatDateDetailPipe } from '../../../components/teammates-common/format-date-detail.pipe';
import { NotificationStyleDescriptionPipe } from
'../../../components/teammates-common/notification-style-description.pipe';

@Component({
  selector: 'tm-notifications-table',
  templateUrl: './notifications-table.component.html',
  styleUrls: ['./notifications-table.component.scss'],
})

export class NotificationsTableComponent implements OnInit {
  SortBy = SortBy;
  SortOrder = SortOrder;
  NotificationsTableHeaderColorScheme = NotificationsTableHeaderColorScheme;

  @Input()
  guessTimezone = 'UTC';

  @Input()
  notifications: NotificationsTableRowModel[] = [];

  @Input()
  notificationsTableRowModelsSortBy = SortBy.NOTIFICATION_CREATE_TIME;

  @Input()
  notificationsTableRowModelsSortOrder = SortOrder.DESC;

  @Input()
  tableSortBy: SortBy = SortBy.NONE;

  @Input()
  tableSortOrder: SortOrder = SortOrder.ASC;

  @Input() set notificationsTableRowModels(notificationRowModels: NotificationsTableRowModel[]) {
    this.notifications = notificationRowModels;
    this.setRowData();
  }

  @Input()
  headerColorScheme = SortableTableHeaderColorScheme.BLUE;

  @Output()
  sortNotificationsTableRowModelsEvent: EventEmitter<SortableEvent> = new EventEmitter();

  @Output()
  addNotificationEvent: EventEmitter<void> = new EventEmitter();

  @Output()
  deleteNotificationEvent: EventEmitter<string> = new EventEmitter();

  @Output()
  loadNotificationEditFormEvent: EventEmitter<Notification> = new EventEmitter();

  notificationStyleDescriptionPipe = new NotificationStyleDescriptionPipe();
  dateDetailPipe = new FormatDateDetailPipe(this.timeZoneService);
  dateBriefPipe = new FormatDateBriefPipe(this.timeZoneService);
  rowsData : SortableTableCellData[][] = [];
  columnsData : ColumnData[] = [];

  backgroundColor: string = 'background-color: #198754;';

  constructor(private simpleModalService: SimpleModalService, private timeZoneService: TimezoneService) { }

  ngOnInit(): void {
    this.setRowData();
    this.setColumnData();
  }

  setColumnData(): void {
    this.columnsData = [
        {
            header: 'Title',
            sortBy: SortBy.NOTIFICATION_TITLE,
            headerClass: 'sort-by-title',
        },
        {
            header: 'Start Time',
            sortBy: SortBy.NOTIFICATION_START_TIME,
            headerClass: 'sort-by-start-time',
        },
        {
            header: 'End Time',
            sortBy: SortBy.NOTIFICATION_END_TIME,
            headerClass: 'sort-by-end-time',
        },
        {
            header: 'Target User',
            sortBy: SortBy.NOTIFICATION_TARGET_USER,
            headerClass: 'sort-by-target-user',
        },
        {
            header: 'Style',
            sortBy: SortBy.NOTIFICATION_STYLE,
            headerClass: 'sort-by-style',

        },
        {
          header: 'Creation Time',
          sortBy: SortBy.NOTIFICATION_CREATE_TIME,
          headerClass: 'sort-by-create-time',
        },
        {
            header: 'Action(s)',
        },
    ];
  }

  setRowData(): void {
    this.rowsData = this.notifications.map((rowModel) => {
      const rowData: SortableTableCellData[] = [
        {
          value: rowModel.notification.title,
          style: rowModel.isHighlighted ? this.backgroundColor : '',
        },
        {
          value: rowModel.notification.startTimestamp,
          displayValue: this.dateBriefPipe.transform(rowModel.notification.startTimestamp, this.guessTimezone),
          tooltip: this.dateDetailPipe.transform(rowModel.notification.startTimestamp, this.guessTimezone),
          style: rowModel.isHighlighted ? this.backgroundColor : '',
          class: 'ngb-tooltip-class',
        },
        {
          value: rowModel.notification.endTimestamp,
          displayValue: this.dateBriefPipe.transform(rowModel.notification.endTimestamp, this.guessTimezone),
          tooltip: this.dateDetailPipe.transform(rowModel.notification.endTimestamp, this.guessTimezone),
          style: rowModel.isHighlighted ? this.backgroundColor : '',
          class: 'ngb-tooltip-class',
        },
        {
          value: rowModel.notification.targetUser,
          displayValue: rowModel.notification.targetUser,
          style: rowModel.isHighlighted ? this.backgroundColor : '',

        },
        {
          value: rowModel.notification.style,
          displayValue: this.notificationStyleDescriptionPipe.transform(rowModel.notification.style),
          style: rowModel.isHighlighted ? this.backgroundColor : '',

        },
        {
          value: rowModel.notification.createdAt,
          displayValue: this.dateBriefPipe.transform(rowModel.notification.createdAt, this.guessTimezone),
          tooltip: this.dateDetailPipe.transform(rowModel.notification.createdAt, this.guessTimezone),
          style: rowModel.isHighlighted ? this.backgroundColor : '',
          class: 'ngb-tooltip-class',
        },
          this.createActionsCell(rowModel),
      ];

      return rowData;
    });
  }

  createActionsCell(rowModel: NotificationsTableRowModel): SortableTableCellData {
    const actionsCell: SortableTableCellData = {
      style: rowModel.isHighlighted ? this.backgroundColor : '',
      customComponent: {
      component: CellWithActionsComponent,
      componentData: (idx: number) => ({
        idx,
        notificationId: rowModel.notification.notificationId,
        deleteNotification: () => this.deleteNotification(rowModel.notification.notificationId,
          rowModel.notification.title),
        editNotification: () => this.loadNotificationEditForm(rowModel.notification),
      }),
      },
    };

    return actionsCell;
  }

  /**
   * Sorts the list of feedback session row.
   */
  sortNotificationsTableRowModels(event: { sortBy: SortBy, sortOrder: SortOrder }): void {
    this.sortNotificationsTableRowModelsEvent.emit(event);
  }

  getAriaSort(by: SortBy): string {
    if (by !== this.notificationsTableRowModelsSortBy) {
      return 'none';
    }
    return this.notificationsTableRowModelsSortOrder === SortOrder.ASC ? 'ascending' : 'descending';
  }

  /**
   * Deletes a notification based on its ID.
   */
  deleteNotification(notificationId: string, title: string): void {
    const modalRef = this.simpleModalService.openConfirmationModal(
      'Confirm your action',
      SimpleModalType.DANGER,
      `Do you want to delete this notification (titled "${title}") permanently? This action will not be reversible.`,
    );
    modalRef.result.then(() => this.deleteNotificationEvent.emit(notificationId));
  }

  /**
   * Loads the notification edit form.
   */
  loadNotificationEditForm(notification: Notification): void {
    this.loadNotificationEditFormEvent.emit(notification);
  }
}
