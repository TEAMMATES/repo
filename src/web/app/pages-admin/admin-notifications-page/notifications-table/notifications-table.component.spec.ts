import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AdminNotificationsPageModule } from '../admin-notifications-page.module';

import { NotificationsTableComponent } from './notifications-table.component';

describe('NotificationsTableComponent', () => {
  let component: NotificationsTableComponent;
  let fixture: ComponentFixture<NotificationsTableComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        AdminNotificationsPageModule,
        HttpClientTestingModule,
      ],
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(NotificationsTableComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
