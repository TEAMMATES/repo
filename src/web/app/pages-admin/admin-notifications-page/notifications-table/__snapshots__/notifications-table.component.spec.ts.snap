// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsTableComponent should snap like in notification page with 2 notifications sorted by create time 1`] = `
<tm-notifications-table
  NotificationsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  addNotificationEvent={[Function EventEmitter_]}
  backgroundColor={[Function String]}
  columnsData={[Function Array]}
  dateBriefPipe={[Function FormatDateBriefPipe]}
  dateDetailPipe={[Function FormatDateDetailPipe]}
  deleteNotificationEvent={[Function EventEmitter_]}
  guessTimezone={[Function String]}
  headerColorScheme="0"
  loadNotificationEditFormEvent={[Function EventEmitter_]}
  notificationStyleDescriptionPipe={[Function NotificationStyleDescriptionPipe]}
  notifications={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  rowsData={[Function Array]}
  simpleModalService={[Function SimpleModalService]}
  sortNotificationsTableRowModelsEvent={[Function EventEmitter_]}
  tableSortBy="0"
  tableSortOrder={[Function Number]}
  timeZoneService={[Function TimezoneService]}
>
  <div>
    <tm-sortable-table>
      <table
        class="table table-bordered table-striped"
      >
        <thead>
          <tr
            class="bg-primary text-white"
          >
            <th
              aria-sort="none"
              class="sort-by-title sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Title
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-start-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Start Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-end-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  End Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-target-user sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Target User
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-style sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Style
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-create-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Creation Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              style="text-align: start;"
            >
              <span
                container="body"
              >
                Action(s)
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              valid title 1
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span>
                INSTRUCTOR
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span>
                Success (green)
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <ndc-dynamic />
              <tm-group-buttons>
                <div>
                  <button
                    class="btn btn-light btn-sm"
                    style="margin-right: 6px;"
                    type="button"
                  >
                     Edit 
                  </button>
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                     Delete 
                  </button>
                </div>
              </tm-group-buttons>
            </td>
          </tr>
          <tr>
            <td>
              valid title 2
            </td>
            <td>
              <span
                class="ngb-tooltip-class"
              >
                15 Dec 9:30 AM
              </span>
            </td>
            <td>
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td>
              <span>
                GENERAL
              </span>
            </td>
            <td>
              <span>
                Danger (red)
              </span>
            </td>
            <td>
              <span
                class="ngb-tooltip-class"
              >
                15 Nov 9:30 AM
              </span>
            </td>
            <td>
              <ndc-dynamic />
              <tm-group-buttons>
                <div>
                  <button
                    class="btn btn-light btn-sm"
                    style="margin-right: 6px;"
                    type="button"
                  >
                     Edit 
                  </button>
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                     Delete 
                  </button>
                </div>
              </tm-group-buttons>
            </td>
          </tr>
        </tbody>
      </table>
    </tm-sortable-table>
  </div>
</tm-notifications-table>
`;

exports[`NotificationsTableComponent should snap like in notification page with 2 notifications sorted by start time 1`] = `
<tm-notifications-table
  NotificationsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  addNotificationEvent={[Function EventEmitter_]}
  backgroundColor={[Function String]}
  columnsData={[Function Array]}
  dateBriefPipe={[Function FormatDateBriefPipe]}
  dateDetailPipe={[Function FormatDateDetailPipe]}
  deleteNotificationEvent={[Function EventEmitter_]}
  guessTimezone={[Function String]}
  headerColorScheme="0"
  loadNotificationEditFormEvent={[Function EventEmitter_]}
  notificationStyleDescriptionPipe={[Function NotificationStyleDescriptionPipe]}
  notifications={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  rowsData={[Function Array]}
  simpleModalService={[Function SimpleModalService]}
  sortNotificationsTableRowModelsEvent={[Function EventEmitter_]}
  tableSortBy="0"
  tableSortOrder={[Function Number]}
  timeZoneService={[Function TimezoneService]}
>
  <div>
    <tm-sortable-table>
      <table
        class="table table-bordered table-striped"
      >
        <thead>
          <tr
            class="bg-primary text-white"
          >
            <th
              aria-sort="none"
              class="sort-by-title sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Title
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-start-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Start Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-end-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  End Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-target-user sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Target User
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-style sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Style
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-create-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Creation Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              style="text-align: start;"
            >
              <span
                container="body"
              >
                Action(s)
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              valid title 1
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span>
                INSTRUCTOR
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span>
                Success (green)
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td
              style="background-color: rgb(25, 135, 84);"
            >
              <ndc-dynamic />
              <tm-group-buttons>
                <div>
                  <button
                    class="btn btn-light btn-sm"
                    style="margin-right: 6px;"
                    type="button"
                  >
                     Edit 
                  </button>
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                     Delete 
                  </button>
                </div>
              </tm-group-buttons>
            </td>
          </tr>
          <tr>
            <td>
              valid title 2
            </td>
            <td>
              <span
                class="ngb-tooltip-class"
              >
                15 Dec 9:30 AM
              </span>
            </td>
            <td>
              <span
                class="ngb-tooltip-class"
              >
                15 Sep 9:30 AM
              </span>
            </td>
            <td>
              <span>
                GENERAL
              </span>
            </td>
            <td>
              <span>
                Danger (red)
              </span>
            </td>
            <td>
              <span
                class="ngb-tooltip-class"
              >
                15 Nov 9:30 AM
              </span>
            </td>
            <td>
              <ndc-dynamic />
              <tm-group-buttons>
                <div>
                  <button
                    class="btn btn-light btn-sm"
                    style="margin-right: 6px;"
                    type="button"
                  >
                     Edit 
                  </button>
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                     Delete 
                  </button>
                </div>
              </tm-group-buttons>
            </td>
          </tr>
        </tbody>
      </table>
    </tm-sortable-table>
  </div>
</tm-notifications-table>
`;

exports[`NotificationsTableComponent should snap with default fields 1`] = `
<tm-notifications-table
  NotificationsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  addNotificationEvent={[Function EventEmitter_]}
  backgroundColor={[Function String]}
  columnsData={[Function Array]}
  dateBriefPipe={[Function FormatDateBriefPipe]}
  dateDetailPipe={[Function FormatDateDetailPipe]}
  deleteNotificationEvent={[Function EventEmitter_]}
  guessTimezone={[Function String]}
  headerColorScheme="0"
  loadNotificationEditFormEvent={[Function EventEmitter_]}
  notificationStyleDescriptionPipe={[Function NotificationStyleDescriptionPipe]}
  notifications={[Function Array]}
  notificationsTableRowModelsSortBy={[Function Number]}
  notificationsTableRowModelsSortOrder="0"
  rowsData={[Function Array]}
  simpleModalService={[Function SimpleModalService]}
  sortNotificationsTableRowModelsEvent={[Function EventEmitter_]}
  tableSortBy="0"
  tableSortOrder={[Function Number]}
  timeZoneService={[Function TimezoneService]}
>
  <div>
    <tm-sortable-table>
      <table
        class="table table-bordered table-striped"
      >
        <thead>
          <tr
            class="bg-primary text-white"
          >
            <th
              aria-sort="none"
              class="sort-by-title sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Title
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-start-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Start Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-end-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  End Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-target-user sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Target User
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-style sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Style
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              class="sort-by-create-time sortable-header"
              style="text-align: start;"
            >
              <button
                class="d-lg-flex align-items-center gap-lg-1 d-inline"
              >
                <span
                  container="body"
                >
                  Creation Time
                </span>
                <span
                  aria-hidden="true"
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </button>
            </th>
            <th
              aria-sort="none"
              style="text-align: start;"
            >
              <span
                container="body"
              >
                Action(s)
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
        </tbody>
      </table>
    </tm-sortable-table>
  </div>
</tm-notifications-table>
`;
